### PR TITLE
Simplify + combine breach classification methods

### PR DESCRIPTION
**Pull request by kjdchapman**
This PR is a draft because it's just explaining how these methods _could_ look, without actually wiring them up.

We're working on the SQL that sits underneath them now.

### CODE REVIEW

File: lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb [hunk 1]
```diff
@@ -70,6 +73,23 @@ def update_court_outcome_action?
           @criteria.court_outcome.blank?
         end
 
+        def breach_letter_action
+          return unless @criteria.most_recent_agreement.present?
+          return unless @criteria.most_recent_agreement[:start_date].present?
+          return unless @criteria.most_recent_agreement[:breached]
+
+          return :send_informal_agreement_breach_letter_new unless @criteria.courtdate.present?
```
Comment (richrace) [inline lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb]:
```suggestion
          return :send_informal_agreement_breach_letter_new if @criteria.courtdate.blank?
```

https://www.rubydoc.info/gems/rubocop/0.53.0/RuboCop/Cop/Rails/Blank

File: lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb [hunk 2]
```diff
@@ -70,6 +73,23 @@ def update_court_outcome_action?
...
+
+          court_date_after_agreement = @criteria.courtdate > @criteria.most_recent_agreement[:start_date]
+          court_date_long_before_agreement = @criteria.courtdate + 3.months < @criteria.most_recent_agreement[:start_date]
+
+          if court_date_after_agreement || court_date_long_before_agreement
```
Comment (richrace) [inline lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb]: `court_date_long_before_agreement` not a fan of `long` in this variable name.

File: lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb [hunk 3]
```diff
@@ -70,6 +73,23 @@ def update_court_outcome_action?
           @criteria.court_outcome.blank?
         end
 
+        def breach_letter_action
+          return unless @criteria.most_recent_agreement.present?
```
Comment (richrace) [inline lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb]:
```suggestion
          return if @criteria.most_recent_agreement.blank?
```

File: lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb [hunk 4]
```diff
@@ -70,6 +73,23 @@ def update_court_outcome_action?
...
+          return unless @criteria.most_recent_agreement[:start_date].present?
```
Comment (richrace) [inline lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb]:
```suggestion
          return if @criteria.most_recent_agreement[:start_date].blank?
```

File: lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb (hunk 2 above)
Comment (richrace) [reply lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb]: So it's an informal agreement if the court date is after the start date or if the start date happens a "long" time after the court date.

Seems to make sense.

Comment (kjdchapman) [reply lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb]: Yeah I didn't love it either, open to suggestions!

Comment (richrace) [reply lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb]: `agreement_months_before_court_date`?

**Comment by kjdchapman** [comment]
This PR now includes the SQL changes to get the new data, and wires up both sides.

There's some dead code that I want to remove, but I want to release this ASAP, so I'm happy to do that in a later piece of work.

File: lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb [hunk 5]
```diff
@@ -53,35 +53,28 @@ def review_failed_letter?
...
-          return false if @criteria.balance >= @criteria.expected_balance
-          return false unless @criteria.court_outcome.in?(active_agreement_court_outcomes)
-          return false unless @criteria.last_communication_action.in?(valid_actions_for_court_agreement_breach_letter_to_progress)
-          true
+        def breach_letter_action
+          return if @criteria.most_recent_agreement.blank?
+          return if @criteria.most_recent_agreement[:start_date].blank?
+          return unless @criteria.most_recent_agreement[:breached]
+
+          return :send_informal_agreement_breach_letter if @criteria.courtdate.blank?
```
Comment (kjdchapman) [inline lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb]: I don't love this early return, but I couldn't see a nicer way to do it without modifying the "main" method of this class a bunch.

File: lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb [hunk 6]
```diff
@@ -53,35 +53,28 @@ def review_failed_letter?
...
-          return false if @criteria.latest_active_agreement_date <= @criteria.courtdate
-          return false if @criteria.breach_agreement_date.present? && @criteria.breach_agreement_date + 3.days > Date.today
-          return false if @criteria.balance >= @criteria.expected_balance
-          return false unless @criteria.court_outcome.in?(active_agreement_court_outcomes)
-          return false unless @criteria.last_communication_action.in?(valid_actions_for_court_agreement_breach_letter_to_progress)
-          true
+        def breach_letter_action
+          return if @criteria.most_recent_agreement.blank?
+          return if @criteria.most_recent_agreement[:start_date].blank?
+          return unless @criteria.most_recent_agreement[:breached]
```
Comment (richrace) [inline lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb]: Should we have a variable or explicitly return `nil` here?

I had to double check what this function should be returning.

**Pull request merged**